### ISSUE TITLE
multiple aws tags

### ISSUE THREAD

**Issue by daroga0002** [hazelcast/hazelcast-aws#202]
Does there is possibility to user multiple AWS tags for discovery process?

In huge scale where we have multiple `customer` environments shared across one VPC  (seperate ECS clusters) mixed with multiple other EC2 instances it is hard to define one particular tag for filtering. 

For example we using following tagging schema for ECS cluster:
```
    "customer" = *****
    "module"   = *****
    "customer" = *****
    "Cluster"  = *****
    "module"   = "ecs"
```

on the end none of this tag is exclusive. Currently we are trying to workaround this by adding next tag with unque value across whole infrastructure, but I think it should be quite beneficial to add support for multiple tags inside this package.

**Comment by leszko** [comment]
Currently, it's not possible to use multiple tags. But I like this feature. We could have tags separated by a comma and then take them with AND condition.
```
aws:
  tag-key: first-tag-key,second-tag-key
  tag-value: first-tag-value,second-tag-value
```

Feel free to submit a PR if you want to contribute.

**Comment by daroga0002** [comment]
Unfortunately I am not developer so I will be not able to create PR for this, but maybe somebody from community will use this idea

**Comment by ajurasz** [comment]
I would like to work on adding multi-tag support to `hazelcast-aws`. After analyzing the code base, I counted three different strategies for filtering EC2 instances using tags combination:

1) `key and value` strategy - filters EC2 instances by tag key and it's value
2) `key` strategy - filters EC2 instances only by tag key (instances where the given tag key is used and ignoring its value)
3) `value` strategy - filters EC2 instances only by tag value (instances where the given tag value is used and ignoring its key)

I assume we would like to maintain support for these strategies and don't break configuration compatibility. Below is my suggestion on how it could work:

1) `key and value` strategy - this will be activated only for tags where there will be a key and a value.

```yml
aws:
  tag-key: keyA,keyB
  tag-value: valueA,valueB 
```

Resulting tag entries `[(keyA,valueA), (keyB,valueB)]`.

2) `key` strategy - keys for which there are no corresponding values.

```yml
aws:
  tag-key: keyA,keyB,keyC,keyD
  tag-value: valueA,valueB
```

Resulting tag entries `[(keyA,valueA), (keyB,valueB), (keyC,), (keyD,)]`. 

`keyA` and `keyB` will use `key and value` strategy.

`keyC` and `keyD` will use `key` strategy.

3) `value` strategy - values for which there are no corresponding keys.

```yml
aws:
  tag-key: keyA,keyB
  tag-value: valueA,valueB,valueC,valueD
```

Resulting tag entries `[(keyA,valueA), (keyB,valueB), (,valueC), (,valueD)]`. 

`keyA` and `keyB` will use `key and value` strategy.

`valueC` and `valueD` will use `value` strategy.

@leszko does it look as a valid approach?

**Comment by daroga0002** [comment]
I think scenario 3 is not valid as AWS doesnt allow you create tag without key name (what in general make sense):
![image](https://user-images.githubusercontent.com/20227477/105676791-4d2c3380-5eeb-11eb-8015-63bb62bd1a8d.png)

Also next nice to have enhancement can be adding comparing value against wildcard what also should be beneficial for most of companies. 

Some example use case scenario for wildcards:
I have 3 instances with tag `Name` and values:

* hazelcast-1
* hazelcast-2
* hazelcast-3

So adding wildcard support I can cover this scenario via some setup like (in sure it require some escape character for `real` * which is also valid inside tag value):

```bash
aws:
  tag-key: Name
  tag-value: hazelcast*
```

**Comment by leszko** [comment]
Thanks @ajurasz that you want to take care of this feature! And thanks @daroga0002 for the comments.

So first, I think the approach you described is correct. We exclude point 3 (as commented by @daroga0002). And points 1 and 2 are correct as you described them. We extend the functionality and keep the backward-compatiblity.

Concerning the wildcard `*`, it would be a nice feature as well. Can be done as a separate PR / GH Issue (or in the same one if it's trivial to implement).

**Comment by ajurasz** [comment]
@daroga0002  you are right, it's not possible to create tag without the name. What I wanted to express by these scenarios, was to show how can we  **filter** EC2 instances to match provided criteria (`tag-key` and/or `tag-value`).

I did try to limit returned instances from EC2 API while using just `tag-value` and it worked as expected:

```
https://ec2.amazonaws.com/?Action=DescribeInstances&Version=2016-11-15&Filter.1.Name=tag-value&Filter.1.Value.1=ValueA
```

It finds all instances were `tag-value=ValueA` (this does not care under which `tag-key` this value is).


I think I have got everything that's needed to start working on implementation.

**Comment by mtyazici** [comment]
Closing this issue since it is resolved.

**Comment by mtyazici** [comment]
Hi @ajurasz ,

Your PR is released in hazelcast-aws:3.4, thanks again for the contribution!

**Comment by ahoehma** [comment]
Hi all.

what about "OR" ?

I have a scenario where I have two different infrastructures and I would like to run my service on both without breaking backward compatibility. So I need a tag for the old world and tag for the new world ... unfortunately I have only ONE place to define my properties for the service :/ 

on the "old" infra I would like to keep the tag=values which there existing (foo=bar)
but on the "new" infra I would like to run my service with (foo2=blub)

So my idea is to have a props like this  
tag=foo,foo2
value = bar,blub
comb=or
